src/IssueViz.Security/Taint/TaintStore.cs

```diff
@@ -39,21 +40,85 @@ internal interface ITaintStore : IIssuesStore
         /// Returns additional analysis information for the existing visualizations in the store.
         /// </summary>
         AnalysisInformation GetAnalysisInformation();
+
+        /// <summary>
+        /// Add the given issue to the existing list of visualizations.
+        /// If <see cref="GetAnalysisInformation"/> is null, the operation is ignored.
+        /// </summary>
+        void Add(IAnalysisIssueVisualization issueVisualization);
+
+        /// <summary>
+        /// Removes an issue with the given key from the existing list of visualizations.
+        /// If no matching issue is found, the operation is ignored.
+        /// </summary>
+        void Remove(string issueKey);
     }
 
     [Export(typeof(ITaintStore))]
     [Export(typeof(IIssuesStore))]
     [PartCreationPolicy(CreationPolicy.Shared)]
     internal sealed class TaintStore : ITaintStore
     {
-        private IAnalysisIssueVisualization[] taintVulnerabilities = Array.Empty<IAnalysisIssueVisualization>();
+        public event EventHandler<IssuesChangedEventArgs> IssuesChanged;
+
+        private static readonly object Locker = new object();
+
+        private List<IAnalysisIssueVisualization> taintVulnerabilities = Array.Empty<IAnalysisIssueVisualization>().ToList();
         private AnalysisInformation analysisInformation;
 
-        public IReadOnlyCollection<IAnalysisIssueVisualization> GetAll() => taintVulnerabilities;
-        
+        public IReadOnlyCollection<IAnalysisIssueVisualization> GetAll()
+        {
+            lock (Locker)
```

Comment:
I don't think the lock here provides any protection; even though the return type is `IReadOnlyCollection<>`, we are returning the real `List` object, so if we modify later in this class any consumers will see the modified list.

Somewhat surprisingly, even calling `.AsReadOnly()` returns a copy that can be modified:
```
var list = new List<string>() { "1", "2" };
IReadOnlyCollection<string> readOnly = list.AsReadOnly();
Console.WriteLine(($"readOnly count: {readOnly.Count}")); // Writes "2"
list.Add("3");
Console.WriteLine(($"readOnly count: {readOnly.Count}")); // Writes "3"
```

If we really want to return a fixed list we should explicitly copy the list and return the copy.

Reply:
.ToList() also works :-)